src/workerd/api/util.c++

```diff
@@ -284,4 +285,70 @@ double dateNow() {
   return 0.0;
 }
 
+kj::Maybe<jsg::V8Ref<v8::Object>> cloneRequestCf(
+    jsg::Lock& js,
+    kj::Maybe<jsg::V8Ref<v8::Object>> maybeCf) {
+  KJ_IF_MAYBE(cf, maybeCf) {
+    // In case the cf object has a logging proxy, we want to make sure
+    // the logging is not triggered here when the object is cloned.
+    NoRequestCfProxyLoggingScope noLoggingScope;
+    auto cloned = cf->deepClone(js);
+    maybeWrapBotManagement(js.v8Isolate, cloned.getHandle(js));
+    return kj::mv(cloned);
+  }
+  return nullptr;
+}
+
+void maybeWrapBotManagement(v8::Isolate* isolate, v8::Local<v8::Object> handle) {
+  auto context = isolate->GetCurrentContext();
+  auto botManagement = jsg::v8StrIntern(isolate, "botManagement");
+  v8::Local<v8::Value> maybeBotManagement = jsg::check(handle->Get(context, botManagement));
+  // If the botManagement field exists, we replace its value here with a Proxy object that
+  // will log the first time any of its properties are accessed. Logging will occur only
+  // once per worker instance.
+  //
+  // Replacing the value with a proxy rather than setting an accessor for botManagement on
+  // the request.cf object itself avoids false positives when someone is simply iterating
+  // over the fields of request.cf without actually using them. It also allows us to avoid
+  // having to create accessors or a class to intercept every individual property on the
+  // botManagement object.
+  if (maybeBotManagement->IsObject() && !maybeBotManagement->IsProxy()) {
+    v8::Local<v8::Object> bmObj = maybeBotManagement.As<v8::Object>();
+    auto& js = jsg::Lock::from(isolate);
+
+    // Create the Proxy handler exactly once per global context and cache it using a private
+    // property on the global itself. The handler itself maintains no state so it is safe to
+    // use it over and over for all requests.
+    auto sym = v8::Private::ForApi(isolate, jsg::v8StrIntern(isolate, "loggingProxyHandler"));
+    auto handler = jsg::check(context->Global()->GetPrivate(context, sym));
+    if (handler->IsUndefined()) {
+      handler = v8::Object::New(isolate);
+      jsg::check(handler.As<v8::Object>()->Set(context, jsg::v8StrIntern(isolate, "get"),
+          js.wrapReturningFunction(context,
+              [](jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
+        if (IoContext::hasCurrent()) {
+          if (!NoRequestCfProxyLoggingScope::isActive()) {
+            IoContext::current().getMetrics().logBotManagementUse();
```

Comment:
Should we guard this with `IoContext::hasCurrent()`? It would take some extreme contortions to trigger this outside of a request scope, but I think it could be done with dynamic imports.

Reply:
Not sure what you're asking since there's already a guard for `hasCurrent()` there. If you're asking if we really need to, I think so, if someone happened to store that `request.cf` at the global scope and access it from a dynamic import or global scope queueMicrotask, then we could run into issues.

Reply:
You're right. I'm blind.

Reply:
Yeah wow you caught me, I obviously didn't read the immediately surrounding code, which features the check I was asking for two lines above, and a long comment about it below. :facepalm: